File: cc/hybrid/hpke_private_key.cc
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

#include "tink/hybrid/hpke_private_key.h"

#include <memory>

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"
#include "absl/strings/string_view.h"
#include "tink/internal/bn_util.h"
#ifdef OPENSSL_IS_BORINGSSL
#include "openssl/base.h"
#include "openssl/ec_key.h"
#else
#include "openssl/ec.h"
#endif
#include "tink/hybrid/hpke_parameters.h"
#include "tink/hybrid/hpke_public_key.h"
#include "tink/insecure_secret_key_access.h"
#include "tink/internal/ec_util.h"
#include "tink/internal/err_util.h"
#include "tink/internal/ssl_unique_ptr.h"
#include "tink/key.h"
#include "tink/partial_key_access_token.h"
#include "tink/restricted_data.h"
#include "tink/subtle/common_enums.h"
#include "tink/util/secret_data.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"

namespace crypto {
namespace tink {
namespace {

util::StatusOr<subtle::EllipticCurveType> CurveTypeFromKemId(
    HpkeParameters::KemId kem_id) {
  switch (kem_id) {
    case HpkeParameters::KemId::kDhkemP256HkdfSha256:
      return subtle::EllipticCurveType::NIST_P256;
    case HpkeParameters::KemId::kDhkemP384HkdfSha384:
      return subtle::EllipticCurveType::NIST_P384;
    case HpkeParameters::KemId::kDhkemP521HkdfSha512:
      return subtle::EllipticCurveType::NIST_P521;
    case HpkeParameters::KemId::kDhkemX25519HkdfSha256:
      return subtle::EllipticCurveType::CURVE25519;
    default:
      return util::Status(absl::StatusCode::kInvalidArgument,
                          absl::StrCat("Unknown KEM ID: ", kem_id));
  }
}

util::Status ValidatePrivateKeyLength(HpkeParameters::KemId kem_id,
                                      int length) {
  int expected_length;
  switch (kem_id) {
    // Key lengths from 'Nsk' column in
    // https://www.rfc-editor.org/rfc/rfc9180.html#table-2.
    case HpkeParameters::KemId::kDhkemP256HkdfSha256:
      expected_length = 32;
      break;
    case HpkeParameters::KemId::kDhkemP384HkdfSha384:
      expected_length = 48;
      break;
    case HpkeParameters::KemId::kDhkemP521HkdfSha512:
      expected_length = 66;
      break;
    case HpkeParameters::KemId::kDhkemX25519HkdfSha256:
      expected_length = 32;
      break;
    default:
      return util::Status(absl::StatusCode::kInvalidArgument,
                          absl::StrCat("Unknown KEM ID: ", kem_id));
  }

  // Validate key length.
  if (expected_length != length) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        absl::StrFormat(
            "Invalid private key length for KEM %d (expected %d, got %d)",
            kem_id, expected_length, length));
  }

  return util::OkStatus();
}

bool IsNistKem(HpkeParameters::KemId kem_id) {
  return kem_id == HpkeParameters::KemId::kDhkemP256HkdfSha256 ||
         kem_id == HpkeParameters::KemId::kDhkemP384HkdfSha384 ||
         kem_id == HpkeParameters::KemId::kDhkemP521HkdfSha512;
}

util::Status ValidateNistEcKeyPair(subtle::EllipticCurveType curve,
                                   absl::string_view public_key_bytes,
                                   const util::SecretData& private_key_bytes) {
  // Construct EC_KEY from public and private key bytes.
  util::StatusOr<internal::SslUniquePtr<EC_GROUP>> group =
      internal::EcGroupFromCurveType(curve);
  if (!group.ok()) {
    return group.status();
  }
  internal::SslUniquePtr<EC_KEY> key(EC_KEY_new());
  EC_KEY_set_group(key.get(), group->get());

  util::StatusOr<internal::SslUniquePtr<EC_POINT>> public_key =
      internal::EcPointDecode(curve, subtle::EcPointFormat::UNCOMPRESSED,
                              public_key_bytes);
  if (!public_key.ok()) {
    return public_key.status();
  }

  if (!EC_KEY_set_public_key(key.get(), public_key->get())) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("Invalid public key: ", internal::GetSslErrors()));
  }

  util::StatusOr<internal::SslUniquePtr<BIGNUM>> priv_key =
      internal::StringToBignum(util::SecretDataAsStringView(private_key_bytes));
  if (!priv_key.ok()) {
    return priv_key.status();
  }
  if (!EC_KEY_set_private_key(key.get(), priv_key->get())) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("Invalid private key: ", internal::GetSslErrors()));
  }

  // Check that EC_KEY is valid.
  if (!EC_KEY_check_key(key.get())) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        absl::StrCat("Invalid EC key pair: ", internal::GetSslErrors()));
  }

  return util::OkStatus();
}

util::Status ValidateX25519KeyPair(absl::string_view public_key_bytes,
                                   const util::SecretData& private_key_bytes) {
  util::StatusOr<std::unique_ptr<internal::X25519Key>> x25519_key =
      internal::X25519KeyFromPrivateKey(private_key_bytes);
  if (!x25519_key.ok()) {
    return x25519_key.status();
  }
  auto public_key_bytes_from_private = absl::string_view(
      reinterpret_cast<const char*>((*x25519_key)->public_value),
      internal::X25519KeyPubKeySize());
  if (public_key_bytes != public_key_bytes_from_private) {
    return util::Status(
        absl::StatusCode::kInvalidArgument,
        "X25519 private key does not match the specified X25519 public key.");
  }
  return util::OkStatus();
}

util::Status ValidateKeyPair(const HpkePublicKey& public_key,
                             const RestrictedData& private_key_bytes,
                             PartialKeyAccessToken token) {
  HpkeParameters::KemId kem_id = public_key.GetParameters().GetKemId();
  absl::string_view public_key_bytes = public_key.GetPublicKeyBytes(token);
  util::SecretData secret = util::SecretDataFromStringView(
      private_key_bytes.GetSecret(InsecureSecretKeyAccess::Get()));

  if (IsNistKem(kem_id)) {
    util::StatusOr<subtle::EllipticCurveType> curve =
        CurveTypeFromKemId(kem_id);
    if (!curve.ok()) {
      return curve.status();
    }
    return ValidateNistEcKeyPair(*curve, public_key_bytes, secret);
  }
  return ValidateX25519KeyPair(public_key_bytes, secret);
}

}  // namespace

util::StatusOr<HpkePrivateKey> HpkePrivateKey::Create(
    const HpkePublicKey& public_key, const RestrictedData& private_key_bytes,
    PartialKeyAccessToken token) {
  util::Status key_length_validation = ValidatePrivateKeyLength(
      public_key.GetParameters().GetKemId(), private_key_bytes.size());
  if (!key_length_validation.ok()) {
    return key_length_validation;
  }
  util::Status key_pair_validation =
      ValidateKeyPair(public_key, private_key_bytes, token);
  if (!key_pair_validation.ok()) {
    return key_pair_validation;
  }
  return HpkePrivateKey(public_key, private_key_bytes);
}

bool HpkePrivateKey::operator==(const Key& other) const {
  const HpkePrivateKey* that = dynamic_cast<const HpkePrivateKey*>(&other);
  if (that == nullptr) {
    return false;
  }
  if (public_key_ != that->public_key_) {
    return false;
  }
  return private_key_bytes_ == that->private_key_bytes_;
}

}  // namespace tink
}  // namespace crypto

File: java_src/src/main/java/com/google/crypto/tink/hybrid/HpkePrivateKey.java
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.hybrid;

import com.google.crypto.tink.AccessesPartialKey;
import com.google.crypto.tink.InsecureSecretKeyAccess;
import com.google.crypto.tink.Key;
import com.google.crypto.tink.internal.BigIntegerEncoding;
import com.google.crypto.tink.internal.EllipticCurvesUtil;
import com.google.crypto.tink.subtle.EllipticCurves;
import com.google.crypto.tink.subtle.X25519;
import com.google.crypto.tink.util.SecretBytes;
import com.google.errorprone.annotations.Immutable;
import com.google.errorprone.annotations.RestrictedApi;
import java.math.BigInteger;
import java.security.GeneralSecurityException;
import java.security.spec.ECParameterSpec;
import java.security.spec.ECPoint;
import java.util.Arrays;

/** Representation of the decryption function for an HPKE hybrid encryption primitive. */
@Immutable
public final class HpkePrivateKey extends HybridPrivateKey {
  private final HpkePublicKey publicKey;
  private final SecretBytes privateKeyBytes;

  private HpkePrivateKey(HpkePublicKey publicKey, SecretBytes privateKeyBytes) {
    this.publicKey = publicKey;
    this.privateKeyBytes = privateKeyBytes;
  }

  private static void validatePrivateKeyByteLength(
      HpkeParameters.KemId kemId, SecretBytes privateKeyBytes) throws GeneralSecurityException {
    // Key lengths from 'Nsk' column in https://www.rfc-editor.org/rfc/rfc9180.html#table-2.
    int keyLengthInBytes = privateKeyBytes.size();
    String parameterizedErrorMessage =
        "Encoded private key byte length for " + kemId + " must be %d, not " + keyLengthInBytes;
    if (kemId == HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256) {
      if (keyLengthInBytes != 32) {
        throw new GeneralSecurityException(String.format(parameterizedErrorMessage, 32));
      }
      return;
    }
    if (kemId == HpkeParameters.KemId.DHKEM_P384_HKDF_SHA384) {
      if (keyLengthInBytes != 48) {
        throw new GeneralSecurityException(String.format(parameterizedErrorMessage, 48));
      }
      return;
    }
    if (kemId == HpkeParameters.KemId.DHKEM_P521_HKDF_SHA512) {
      if (keyLengthInBytes != 66) {
        throw new GeneralSecurityException(String.format(parameterizedErrorMessage, 66));
      }
      return;
    }
    if (kemId == HpkeParameters.KemId.DHKEM_X25519_HKDF_SHA256) {
      if (keyLengthInBytes != 32) {
        throw new GeneralSecurityException(String.format(parameterizedErrorMessage, 32));
      }
      return;
    }
    throw new GeneralSecurityException("Unable to validate private key length for " + kemId);
  }

  private static boolean isNistKem(HpkeParameters.KemId kemId) {
    return kemId == HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256
        || kemId == HpkeParameters.KemId.DHKEM_P384_HKDF_SHA384
        || kemId == HpkeParameters.KemId.DHKEM_P521_HKDF_SHA512;
  }

  private static ECParameterSpec getNistCurveParams(HpkeParameters.KemId kemId) {
    if (kemId == HpkeParameters.KemId.DHKEM_P256_HKDF_SHA256) {
      return EllipticCurves.getNistP256Params();
    }
    if (kemId == HpkeParameters.KemId.DHKEM_P384_HKDF_SHA384) {
      return EllipticCurves.getNistP384Params();
    }
    if (kemId == HpkeParameters.KemId.DHKEM_P521_HKDF_SHA512) {
      return EllipticCurves.getNistP521Params();
    }
    throw new IllegalArgumentException("Unable to determine NIST curve params for " + kemId);
  }

  /**
   * Confirms that the private key encoded as {@code privateKeyBytes} corresponds to the public key
   * encoded as {@code publicKeyBytes} given the HPKE {@code kemId}.
   */
  private static void validateKeyPair(
      HpkeParameters.KemId kemId, byte[] publicKeyBytes, byte[] privateKeyBytes)
      throws GeneralSecurityException {
    if (isNistKem(kemId)) {
      ECParameterSpec spec = getNistCurveParams(kemId);
      BigInteger order = spec.getOrder();
      BigInteger privateKey = BigIntegerEncoding.fromUnsignedBigEndianBytes(privateKeyBytes);
      if ((privateKey.signum() <= 0) || (privateKey.compareTo(order) >= 0)) {
        throw new GeneralSecurityException("Invalid private key.");
      }
      ECPoint expectedPoint = EllipticCurvesUtil.multiplyByGenerator(privateKey, spec);
      ECPoint publicPoint =
          EllipticCurves.pointDecode(
              spec.getCurve(), EllipticCurves.PointFormatType.UNCOMPRESSED, publicKeyBytes);
      if (!expectedPoint.equals(publicPoint)) {
        throw new GeneralSecurityException("Invalid private key for public key.");
      }
      return;
    }
    if (kemId == HpkeParameters.KemId.DHKEM_X25519_HKDF_SHA256) {
      byte[] expectedPublicKeyBytes = X25519.publicFromPrivate(privateKeyBytes);
      if (!Arrays.equals(expectedPublicKeyBytes, publicKeyBytes)) {
        throw new GeneralSecurityException("Invalid private key for public key.");
      }
      return;
    }
    throw new IllegalArgumentException("Unable to validate key pair for " + kemId);
  }

  /**
   * Creates a new HPKE private key.
   *
   * @param publicKey Corresponding HPKE public key for this private key
   * @param privateKeyBytes Private key encoded according to
   *     https://www.rfc-editor.org/rfc/rfc9180.html#section-7.1.2
   */
  @AccessesPartialKey
  @RestrictedApi(
      explanation = "Accessing parts of keys can produce unexpected incompatibilities, annotate the function with @AccessesPartialKey",
      link = "https://developers.google.com/tink/design/access_control#accessing_partial_keys",
      allowedOnPath = ".*Test\\.java",
      allowlistAnnotations = {AccessesPartialKey.class})
  public static HpkePrivateKey create(HpkePublicKey publicKey, SecretBytes privateKeyBytes)
      throws GeneralSecurityException {
    if (publicKey == null) {
      throw new GeneralSecurityException(
          "HPKE private key cannot be constructed without an HPKE public key");
    }
    if (privateKeyBytes == null) {
      throw new GeneralSecurityException("HPKE private key cannot be constructed without secret");
    }
    validatePrivateKeyByteLength(publicKey.getParameters().getKemId(), privateKeyBytes);
    validateKeyPair(
        publicKey.getParameters().getKemId(),
        publicKey.getPublicKeyBytes().toByteArray(),
        privateKeyBytes.toByteArray(InsecureSecretKeyAccess.get()));
    return new HpkePrivateKey(publicKey, privateKeyBytes);
  }

  @RestrictedApi(
      explanation = "Accessing parts of keys can produce unexpected incompatibilities, annotate the function with @AccessesPartialKey",
      link = "https://developers.google.com/tink/design/access_control#accessing_partial_keys",
      allowedOnPath = ".*Test\\.java",
      allowlistAnnotations = {AccessesPartialKey.class})
  public SecretBytes getPrivateKeyBytes() {
    return privateKeyBytes;
  }

  @Override
  public HpkeParameters getParameters() {
    return publicKey.getParameters();
  }

  @Override
  public HpkePublicKey getPublicKey() {
    return publicKey;
  }

  @Override
  public boolean equalsKey(Key o) {
    if (!(o instanceof HpkePrivateKey)) {
      return false;
    }
    HpkePrivateKey other = (HpkePrivateKey) o;
    return publicKey.equalsKey(other.publicKey)
        && privateKeyBytes.equalsSecretBytes(other.privateKeyBytes);
  }
}
